### PR TITLE
fix: Hardcode known intermediary thumbprints in github oidc provider

### PR DESCRIPTION
**Pull request by therealdwright**
## Description
Following GitHub's announcement that can be found on their [blog](https://github.blog/changelog/2023-06-27-github-actions-update-on-oidc-integration-with-aws/) there are now multiple thumbprints that are required to integrate AWS with GitHub actions.

We have been running with this change for some time now and wanted to submit upstream in case others are having issues with their GitHub actions workflows assuming AWS roles.

## Motivation and Context
Solves errors where GitHub actions workflows are unable to assume an IAM role due to an invalid thumbprint.

## Breaking Changes
N/A - this change is additive and won't interfere with the current implementation.

## How Has This Been Tested?
- [x] I have updated at least one of the `examples/*` to demonstrate and validate my change(s) (N/A)
- [X] I have tested and validated these changes using one or more of the provided `examples/*` projects
- [X] I have executed `pre-commit run -a` on my pull request

As previously mentioned, we have been testing this change for a while. We vendored this change into our own Terraform workspace and have been running without issue for a while now. During the last couple of weeks, GitHub has changed thumbprints twice and our change detected when this occurred and showed up in a Terraform plan.

### CODE REVIEW

**Review by antonbabenko — CHANGES_REQUESTED** [review]
I don't see how this differs from what we already have in this module.

For me, these two data sources produce the same results:
```hcl
data "tls_certificate" "new" {
  url = "https://token.actions.githubusercontent.com/.well-known/openid-configuration"
}
```

```hcl
data "tls_certificate" "current" {
  url = "https://token.actions.githubusercontent.com"
}
```

```
sha1_fingerprint = [
  "6938fd4d98bab03faadb97b34396831e3780aea1",
  "f879abce0008e4eb126e0097e46620f5aaae26ad",
]
```

`1c58a3a8518e8759bf075b76b750d4f2df264fcd` is missing for me.

As discussed in [this comment](https://github.com/aws-actions/configure-aws-credentials/issues/357#issuecomment-1610172322), I think we should do a call to `distinct(concat(...))` to get a list of at least 3 fingerprints (assuming GH will roll the updates in the future and we don't need to update it again).

Will you be able to update the code accordingly?

**Comment by therealdwright** [comment]
> Every second request to the data source changes

Thank you for bringing this to my attention I thought I was going crazy and getting a bit better stability with the endpoint I provided, but it was wrong. I think I agree that at least in the interim until these fingerprints are stable that hardcoding is the right approach. I will update the PR to use the fingerprints provided by GitHub and leave it to the maintainers to decide if they want a TODO or to merge at all.

**Comment by bryantbiggs** [comment]
Here we can simply wrap it in a `sort()` to make the diff the same - we should NOT hardcode these

**Comment by arianvp** [comment]
> Here we can simply wrap it in a sort() to make the diff the same - we should NOT hardcode these

This is simply not possible.  Github is returning a different intermediate cert on each request. How do you want to not hardcode this?

**Comment by bryantbiggs** [comment]
> This happens for customers who are pinned to a single intermediary thumbprint from the Certificate Authority (CA) of the Actions SSL certificate.

Pinning something that is known to change, and designed to change, is not a viable solution

**Comment by therealdwright** [comment]
> Pinning something that is known to change, and designed to change, is not a viable solution

Let me know if you would like this PR closed. We have pinned our own custom implementation as this provider is not usable in its current form.

The commit message states it's an interim solution.

**Comment by bryantbiggs** [comment]
why is it not usable in its current form? its returning the wrong thumbprints? its returning the incorrect thumbprints?

**Comment by therealdwright** [comment]
> why is it not usable in its current form? its returning the wrong thumbprints? its returning the incorrect thumbprints?

The provider returns different thumbprints on every CI run leading to detected changes on every single Terraform plan. We run CI automation that plans our infrastructure every PR to our CI environment and it's not viable for us to have unrelated changes in every single PR.

**Comment by therealdwright** [comment]
If you would like me to go with a conditional approach where pinned thumbprints are an optional variable I will do this. However, I'm not going to invest any more time until I get some consensus from maintainers on whether or not this is something they want to solve.

**Comment by bryantbiggs** [comment]
while spurious diffs are annoying, I would prefer to have that rather than having a hardcoded value that will change and be incorrect at some point in the future. I think the proper solution here is for GitHub to provide the guidance or changes to alleviate this issue. As Anton mentioned above, there is a whole discussion happening here https://github.com/aws-actions/configure-aws-credentials/issues/357 with no resolution currently

**Comment by flokli** [comment]
@bryantbiggs you misunderstood. Github actions HTTPS endpoint uses either one certificate chain or another certificate chain for different HTTPS connections, it's pure luck which one you end up with.

A single terraform run might end up discovering one fingerprint, while another one will discover the other one. There is simply no way to properly discover this from Terraform in a reliable fashion, see https://github.com/aws-actions/configure-aws-credentials/issues/357#issuecomment-1612005962 for a detailed answer.

The proper fix would be AWS "Internal P82323250" issue, but in the meantime, pinning certs as documented by AWS in https://github.blog/changelog/2023-06-27-github-actions-update-on-oidc-integration-with-aws/ is the officially documented way, so that's what this terraform module should do too.

**Comment by therealdwright** [comment]
> A single terraform run might end up discovering one fingerprint, while another one will discover the other one

I think the maintainers understand the issue they just don't agree with pinning the images or even providing the option to do so. It's, for this reason, we decided to abandon using the module.

**Pull request closed without merging**
